build failure with Go, AWS-LC, and musl

**Image I'm using:**
Latest `develop`.


**Issue or Feature Request:**
AWS-LC is currently built with the GNU toolchain, and after glibc is updated to 2.38, it starts to depend on newer symbols. This causes problems for the current method of building static Go FIPS binaries, since those must be linked with musl which doesn't provide the same symbols:
```
  #14 142.1 + gofips build -ldflags '-compressdwarf=false -linkmode=external -extldflags '\''-Wl,-z,relro -Wl,-z,now -Wl,-Bsymbolic-functions -static-pie'\''' -o fips-static/amazon
-ssm-agent ./core/agent.go ./core/agent_unix.go ./core/agent_parser.go
  #14 149.8 # crypto/internal/boring
  #14 149.8 cgo-gcc-prolog: In function ‘_cgo_15785f09c110_Cfunc__goboringcrypto_BORINGSSL_bcm_power_on_self_test’:
  #14 149.8 cgo-gcc-prolog:141:49: warning: unused variable ‘_cgo_a’ [-Wunused-variable]                                                                                              #14 160.7 # command-line-arguments                                                                                                                                                  #14 160.7 /usr/libexec/go-1.22/pkg/tool/linux_amd64/link: running x86_64-bottlerocket-linux-musl-gcc failed: exit status 1                                                          #14 160.7 /usr/lib/gcc/x86_64-bottlerocket-linux-musl/11/../../../../x86_64-bottlerocket-linux-musl/bin/ld: /tmp/go-link-1440672810/000032.o: in function `BIO_get_fd':             #14 160.7 (.text+0xe84f): undefined reference to `fopen64'                                                                                                                          #14 160.7 /usr/lib/gcc/x86_64-bottlerocket-linux-musl/11/../../../../x86_64-bottlerocket-linux-musl/bin/ld: /tmp/go-link-1440672810/000032.o: in function `BIO_new_file':
  #14 160.7 (.text+0xea44): undefined reference to `fopen64'
  #14 160.7 /usr/lib/gcc/x86_64-bottlerocket-linux-musl/11/../../../../x86_64-bottlerocket-linux-musl/bin/ld: /tmp/go-link-1440672810/000032.o: in function `bn_mul_mont':
  #14 160.7 (.text+0x25d3e8): undefined reference to `__isoc23_sscanf'
  #14 160.7 collect2: error: ld returned 1 exit status
```

I expect the right answer is to provide copies of the AWS-LC syso that are linked with musl, to avoid this and potentially more subtle errors.

--- a/Dockerfile
+++ b/Dockerfile
@@ -598,39 +598,87 @@ COPY ./helpers/aws-lc/* .
 
 # =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=
 
-FROM sdk-go-1.23-prep as sdk-go-1.23-aws-lc-x86_64
+FROM sdk-go-1.23-prep as sdk-go-1.23-aws-lc-gnu-x86_64
 ENV ARCH="x86_64"
-RUN ./build-aws-lc.sh --arch="${ARCH}" --go-dir="${HOME}/sdk-go"
+ENV LIBC="gnu"
+ENV TARGET="${ARCH}-bottlerocket-linux-${LIBC}"
+RUN ./build-aws-lc.sh --arch="${ARCH}" --target="${TARGET}" --go-dir="${HOME}/sdk-go"
 
 # =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=
 
-FROM sdk-go-1.23-prep as sdk-go-1.23-aws-lc-aarch64
+FROM sdk-go-1.23-prep as sdk-go-1.23-aws-lc-gnu-aarch64
 ENV ARCH="aarch64"
-RUN ./build-aws-lc.sh --arch="${ARCH}" --go-dir="${HOME}/sdk-go"
+ENV LIBC="gnu"
+ENV TARGET="${ARCH}-bottlerocket-linux-${LIBC}"
+RUN ./build-aws-lc.sh --arch="${ARCH}" --target="${TARGET}" --go-dir="${HOME}/sdk-go"
 
 # =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=
 
-FROM sdk-go-1.22-prep as sdk-go-1.22-aws-lc-x86_64
+FROM sdk-go-1.23-prep as sdk-go-1.23-aws-lc-musl-x86_64
 ENV ARCH="x86_64"
-RUN ./build-aws-lc.sh --arch="${ARCH}" --go-dir="${HOME}/sdk-go"
+ENV LIBC="musl"
+ENV TARGET="${ARCH}-bottlerocket-linux-${LIBC}"
+RUN ./build-aws-lc.sh --arch="${ARCH}" --target="${TARGET}" --go-dir="${HOME}/sdk-go"
 
 # =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=
 
-FROM sdk-go-1.22-prep as sdk-go-1.22-aws-lc-aarch64
+FROM sdk-go-1.23-prep as sdk-go-1.23-aws-lc-musl-aarch64
 ENV ARCH="aarch64"
-RUN ./build-aws-lc.sh --arch="${ARCH}" --go-dir="${HOME}/sdk-go"
+ENV LIBC="musl"
+ENV TARGET="${ARCH}-bottlerocket-linux-${LIBC}"
+RUN ./build-aws-lc.sh --arch="${ARCH}" --target="${TARGET}" --go-dir="${HOME}/sdk-go"
+
+# =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=
+
+FROM sdk-go-1.22-prep as sdk-go-1.22-aws-lc-gnu-x86_64
+ENV ARCH="x86_64"
+ENV LIBC="gnu"
+ENV TARGET="${ARCH}-bottlerocket-linux-${LIBC}"
+RUN ./build-aws-lc.sh --arch="${ARCH}" --target="${TARGET}" --go-dir="${HOME}/sdk-go"
+
+# =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=
+
+FROM sdk-go-1.22-prep as sdk-go-1.22-aws-lc-gnu-aarch64
+ENV ARCH="aarch64"
+ENV LIBC="gnu"
+ENV TARGET="${ARCH}-bottlerocket-linux-${LIBC}"
+RUN ./build-aws-lc.sh --arch="${ARCH}" --target="${TARGET}" --go-dir="${HOME}/sdk-go"
+
+# =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=
+
+FROM sdk-go-1.22-prep as sdk-go-1.22-aws-lc-musl-x86_64
+ENV ARCH="x86_64"
+ENV LIBC="musl"
+ENV TARGET="${ARCH}-bottlerocket-linux-${LIBC}"
+RUN ./build-aws-lc.sh --arch="${ARCH}" --target="${TARGET}" --go-dir="${HOME}/sdk-go"
+
+# =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=
+
+FROM sdk-go-1.22-prep as sdk-go-1.22-aws-lc-musl-aarch64
+ENV ARCH="aarch64"
+ENV LIBC="musl"
+ENV TARGET="${ARCH}-bottlerocket-linux-${LIBC}"
+RUN ./build-aws-lc.sh --arch="${ARCH}" --target="${TARGET}" --go-dir="${HOME}/sdk-go"
 
 # =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=
 
 FROM sdk-go-1.23-prep as sdk-go-1.23
 
-COPY --from=sdk-go-1.23-aws-lc-x86_64 \
+COPY --from=sdk-go-1.23-aws-lc-gnu-x86_64 \
   /home/builder/aws-lc/build/goboringcrypto_linux_amd64.syso \
   /home/builder/sdk-go/src/crypto/internal/boring/syso/goboringcrypto_linux_amd64.syso
 
-COPY --from=sdk-go-1.23-aws-lc-aarch64 \
+COPY --from=sdk-go-1.23-aws-lc-gnu-aarch64 \
   /home/builder/aws-lc/build/goboringcrypto_linux_arm64.syso \
   /home/builder/sdk-go/src/crypto/internal/boring/syso/goboringcrypto_linux_arm64.syso
+
+COPY --from=sdk-go-1.23-aws-lc-musl-x86_64 \
+  /home/builder/aws-lc/build/goboringcrypto_linux_amd64.syso \
+  /home/builder/sdk-go/src/crypto/internal/boring/syso/goboringcrypto_linux_musl_amd64.syso
+
+COPY --from=sdk-go-1.23-aws-lc-musl-aarch64 \
+  /home/builder/aws-lc/build/goboringcrypto_linux_arm64.syso \
+  /home/builder/sdk-go/src/crypto/internal/boring/syso/goboringcrypto_linux_musl_arm64.syso
 
 COPY ./helpers/go/* ./
 
@@ -641,13 +689,21 @@ RUN ./build-go.sh --go-version=${GO123VER}
 
 FROM sdk-go-1.22-prep as sdk-go-1.22
 
-COPY --from=sdk-go-1.22-aws-lc-x86_64 \
+COPY --from=sdk-go-1.22-aws-lc-gnu-x86_64 \
   /home/builder/aws-lc/build/goboringcrypto_linux_amd64.syso \
   /home/builder/sdk-go/src/crypto/internal/boring/syso/goboringcrypto_linux_amd64.syso
 
-COPY --from=sdk-go-1.22-aws-lc-aarch64 \
+COPY --from=sdk-go-1.22-aws-lc-gnu-aarch64 \
   /home/builder/aws-lc/build/goboringcrypto_linux_arm64.syso \
   /home/builder/sdk-go/src/crypto/internal/boring/syso/goboringcrypto_linux_arm64.syso
+
+COPY --from=sdk-go-1.22-aws-lc-musl-x86_64 \
+  /home/builder/aws-lc/build/goboringcrypto_linux_amd64.syso \
+  /home/builder/sdk-go/src/crypto/internal/boring/syso/goboringcrypto_linux_musl_amd64.syso
+
+COPY --from=sdk-go-1.22-aws-lc-musl-aarch64 \
+  /home/builder/aws-lc/build/goboringcrypto_linux_arm64.syso \
+  /home/builder/sdk-go/src/crypto/internal/boring/syso/goboringcrypto_linux_musl_arm64.syso
 
 COPY ./helpers/go/* ./
 
@@ -1253,6 +1309,18 @@ COPY --chown=0:0 --from=sdk-go-1.22 \
   /home/builder/sdk-go/licenses/ \
   /usr/share/licenses/go-1.22/
 
+# Create Go trees for the different glibc and musl builds of the AWS-LC syso.
+# Sync timestamps to avoid rebuilds of the Go standard library.
+RUN \
+  for v in 1.22 1.23 ; do \
+    find /usr/libexec/go-${v} -type f -exec touch -r /usr/libexec/go-${v}/bin/go {} \+ && \
+    rsync -aq --link-dest=/usr/libexec/go-${v}/ /usr/libexec/go-${v}{,-musl}/ && \
+    rm /usr/libexec/go-${v}/src/crypto/internal/boring/syso/goboringcrypto_linux_musl_{arm,amd}64.syso && \
+    rm /usr/libexec/go-${v}-musl/src/crypto/internal/boring/syso/goboringcrypto_linux_{arm,amd}64.syso && \
+    mv /usr/libexec/go-${v}-musl/src/crypto/internal/boring/syso/goboringcrypto_linux_{musl_,}amd64.syso && \
+    mv /usr/libexec/go-${v}-musl/src/crypto/internal/boring/syso/goboringcrypto_linux_{musl_,}arm64.syso ; \
+  done
+
 # "sdk-rust-tools" has our attribution generation and license scan tools.
 COPY --chown=0:0 --from=sdk-rust-tools /usr/libexec/tools/ /usr/libexec/tools/
 COPY --chown=0:0 --from=sdk-rust-tools /usr/share/licenses/bottlerocket-license-scan/ /usr/share/licenses/bottlerocket-license-scan/
@@ -1363,11 +1431,6 @@ RUN \
 COPY ./wrappers/go/go /usr/bin/go
 COPY ./wrappers/go/gofmt /usr/bin/gofmt
 COPY ./wrappers/go/gofips /usr/bin/gofips
-
-# Add Go programs to $PATH and sync timestamps to avoid rebuilds.
-RUN \
-  find /usr/libexec/go-1.23 -type f -exec touch -r /usr/libexec/go-1.23/bin/go {} \+ && \
-  find /usr/libexec/go-1.22 -type f -exec touch -r /usr/libexec/go-1.22/bin/go {} \+
 
 # Strip and add tools to the path.
 RUN \

--- a/configs/aws-lc/aarch64-bottlerocket-linux-musl.toolchain.cmake
+++ b/configs/aws-lc/aarch64-bottlerocket-linux-musl.toolchain.cmake
@@ -1,0 +1,13 @@
+set(CMAKE_SYSTEM_NAME Linux)
+set(CMAKE_SYSTEM_PROCESSOR aarch64)
+
+set(CMAKE_SYSROOT /aarch64-bottlerocket-linux-musl/sys-root)
+set(CMAKE_C_COMPILER aarch64-bottlerocket-linux-musl-gcc)
+set(CMAKE_C_COMPILER_TARGET aarch64-bottlerocket-linux-musl)
+set(CMAKE_CXX_COMPILER aarch64-bottlerocket-linux-musl-g++)
+set(CMAKE_CXX_COMPILER_TARGET aarch64-bottlerocket-linux-musl)
+
+set(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM NEVER)
+set(CMAKE_FIND_ROOT_PATH_MODE_LIBRARY ONLY)
+set(CMAKE_FIND_ROOT_PATH_MODE_INCLUDE ONLY)
+set(CMAKE_FIND_ROOT_PATH_MODE_PACKAGE ONLY)

--- a/configs/aws-lc/x86_64-bottlerocket-linux-musl.toolchain.cmake
+++ b/configs/aws-lc/x86_64-bottlerocket-linux-musl.toolchain.cmake
@@ -1,0 +1,13 @@
+set(CMAKE_SYSTEM_NAME Linux)
+set(CMAKE_SYSTEM_PROCESSOR x86_64)
+
+set(CMAKE_SYSROOT /x86_64-bottlerocket-linux-musl/sys-root)
+set(CMAKE_C_COMPILER x86_64-bottlerocket-linux-musl-gcc)
+set(CMAKE_C_COMPILER_TARGET x86_64-bottlerocket-linux-musl)
+set(CMAKE_CXX_COMPILER x86_64-bottlerocket-linux-musl-g++)
+set(CMAKE_CXX_COMPILER_TARGET x86_64-bottlerocket-linux-musl)
+
+set(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM NEVER)
+set(CMAKE_FIND_ROOT_PATH_MODE_LIBRARY ONLY)
+set(CMAKE_FIND_ROOT_PATH_MODE_INCLUDE ONLY)
+set(CMAKE_FIND_ROOT_PATH_MODE_PACKAGE ONLY)

--- a/helpers/aws-lc/build-aws-lc.sh
+++ b/helpers/aws-lc/build-aws-lc.sh
@@ -8,13 +8,14 @@ for opt in "$@"; do
    optarg="$(expr "${opt}" : '[^=]*=\(.*\)')"
    case "${opt}" in
       --arch=*) ARCH="${optarg}" ;;
+      --target=*) TARGET="${optarg}" ;;
       --go-dir=*) GODIR="${optarg}" ;;
    esac
 done
 
 ARCH="${ARCH:?}"
 GODIR="${GODIR:?}"
-TARGET="${ARCH}-bottlerocket-linux-gnu"
+TARGET="${TARGET:?}"
 
 # Some of the AWS-LC sources are built with `-O0`. This is not compatible with
 # `-Wp,-D_FORTIFY_SOURCE=2`, which needs at least `-O2`. Add `-DGOBORING` to

--- a/macros/shared
+++ b/macros/shared
@@ -225,6 +225,7 @@ CROSS_CMAKE_TOOLCHAIN_EOF\
   GOPATH="${PWD}/GOPATH" ; export GOPATH ; \
   GOOS="%{_cross_go_os}" ; export GOOS ; \
   GOARCH="%{_cross_go_arch}" ; export GOARCH ; \
+  GOLIBC="%2" ; export GOLIBC ; \
   CC="%1-gcc" ; export CC ; \
   CC_FOR_TARGET="%1-gcc" ; export CC_FOR_TARGET ; \
   CC_FOR_%{_cross_go_os}_%{_cross_go_arch}="%1-gcc" ; export CC_FOR_%{_cross_go_os}_%{_cross_go_arch} ; \
@@ -239,7 +240,7 @@ CROSS_CMAKE_TOOLCHAIN_EOF\
   PKG_CONFIG_ALLOW_CROSS=1 ; export PKG_CONFIG_ALLOW_CROSS ; \
   GOFLAGS="-mod=vendor -a -buildmode=pie"; export GOFLAGS ; \
   GOLDFLAGS_INT="-compressdwarf=false -linkmode=external" ; export GOLDFLAGS_INT ; \
-  GOLDFLAGS_EXT="%{expand:%%%2}" ; export GOLDFLAGS_EXT ; \
+  GOLDFLAGS_EXT="%{expand:%%%3}" ; export GOLDFLAGS_EXT ; \
   GOLDFLAGS="${GOLDFLAGS_INT} -extldflags '${GOLDFLAGS_EXT}'" ; export GOLDFLAGS ; \
   GOPROXY="off"; export GOPROXY ; \
   GOSUMDB="off"; export GOSUMDB ; \
@@ -249,10 +250,10 @@ CROSS_CMAKE_TOOLCHAIN_EOF\
 %_cross_go_extldflags_static %{_cross_ldflags} -static-pie
 
 %set_cross_go_flags \
-  %set_cross_go_env %{_cross_target} _cross_go_extldflags
+  %set_cross_go_env %{_cross_target} gnu _cross_go_extldflags
 
 %set_cross_go_flags_static \
-  %set_cross_go_env %{_cross_triple}-musl _cross_go_extldflags_static
+  %set_cross_go_env %{_cross_triple}-musl musl _cross_go_extldflags_static
 
 %cross_go_setup() \
   mkdir -p GOPATH/src/%2 ; \

--- a/wrappers/go/go
+++ b/wrappers/go/go
@@ -10,4 +10,9 @@ if [ -n "${GO_VERSION:-}" ] ; then
   GO_MAJOR="${GO_VERSION%.*}"
 fi
 
+# prefer the musl flavor if GOLIBC indicates that we should
+if [ "${GOLIBC:-}" = "musl" ] ; then
+  GO_MAJOR+="-musl"
+fi
+
 exec /usr/libexec/go-"${GO_MAJOR}"/bin/go "${@}"

--- a/wrappers/go/gofips
+++ b/wrappers/go/gofips
@@ -10,5 +10,10 @@ if [ -n "${GO_VERSION:-}" ] ; then
   GO_MAJOR="${GO_VERSION%.*}"
 fi
 
+# prefer the musl flavor if GOLIBC indicates that we should
+if [ "${GOLIBC:-}" = "musl" ] ; then
+  GO_MAJOR+="-musl"
+fi
+
 export GOEXPERIMENT=boringcrypto
 exec /usr/libexec/go-"${GO_MAJOR}"/bin/go "${@}"

--- a/wrappers/go/gofmt
+++ b/wrappers/go/gofmt
@@ -10,4 +10,9 @@ if [ -n "${GO_VERSION:-}" ] ; then
   GO_MAJOR="${GO_VERSION%.*}"
 fi
 
+# prefer the musl flavor if GOLIBC indicates that we should
+if [ "${GOLIBC:-}" = "musl" ] ; then
+  GO_MAJOR+="-musl"
+fi
+
 exec /usr/libexec/go-"${GO_MAJOR}"/bin/gofmt "${@}"

